pkg/operator/csi/csidrivercontroller/sync.go

```diff
@@ -0,0 +1,266 @@
+package csidrivercontroller
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	opv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+func (c *CSIDriverController) syncDeployment(spec *opv1.OperatorSpec, status *opv1.OperatorStatus) (*appsv1.Deployment, error) {
+	deploy := c.getExpectedDeployment(spec)
+
+	deploy, _, err := resourceapply.ApplyDeployment(
+		c.kubeClient.AppsV1(),
+		c.eventRecorder,
+		deploy,
+		resourcemerge.ExpectedDeploymentGeneration(deploy, status.Generations))
+	if err != nil {
+		return nil, err
+	}
+
+	return deploy, nil
+}
+
+func (c *CSIDriverController) syncDaemonSet(spec *opv1.OperatorSpec, status *opv1.OperatorStatus) (*appsv1.DaemonSet, error) {
+	daemonSet := c.getExpectedDaemonSet(spec)
+
+	daemonSet, _, err := resourceapply.ApplyDaemonSet(
+		c.kubeClient.AppsV1(),
+		c.eventRecorder,
+		daemonSet,
+		resourcemerge.ExpectedDaemonSetGeneration(daemonSet, status.Generations))
+	if err != nil {
+		return nil, err
+	}
+
+	return daemonSet, nil
+}
+
+func (c *CSIDriverController) syncStatus(
+	meta *metav1.ObjectMeta,
+	status *opv1.OperatorStatus,
+	deployment *appsv1.Deployment,
+	daemonSet *appsv1.DaemonSet) error {
+
+	// Set the last generation change we dealt with
+	status.ObservedGeneration = meta.Generation
+
+	// Node Service is mandatory, so always set set generation in DaemonSet
+	resourcemerge.SetDaemonSetGeneration(&status.Generations, daemonSet)
+
+	// Set number of replicas of DaemonSet
+	if daemonSet.Status.NumberUnavailable == 0 {
+		status.ReadyReplicas = daemonSet.Status.UpdatedNumberScheduled
+	}
+
+	// Controller Service is not mandatory, so optionally set the generation in Deployment
+	if c.controllerManifest != nil {
+		// CSI Controller Service was deployed, set deployment generation
+		resourcemerge.SetDeploymentGeneration(&status.Generations, deployment)
+
+		// Add number of CSI controllers to the number of replicas ready
+		if deployment != nil {
+			if deployment.Status.UnavailableReplicas == 0 && daemonSet.Status.NumberUnavailable == 0 {
+				status.ReadyReplicas += deployment.Status.UpdatedReplicas
+			}
+		}
+	}
+
+	// Finally, set the conditions
+
+	// The operator does not have any prerequisites (at least now)
+	v1helpers.SetOperatorCondition(&status.Conditions,
+		opv1.OperatorCondition{
+			Type:   fmt.Sprintf("%s%s", c.name, opv1.OperatorStatusTypePrereqsSatisfied),
```

Comment:
This may have been an evolutionary dead end. I don't remember using it anytime recently.

Reply:
If that's OK, I'll remove it. I thought it was mandatory